### PR TITLE
controller/compute: sequence external read hold changes

### PR DESCRIPTION
**Pull request by teskje**
Changes to `ReadHold`s on compute collections are now sequenced through the `command_{tx,rx}` channels of the respective `Instance` controllers. This ensures that `Instance`s observe read hold changes always after they have observed the creation of the affected collection, which wasn't the case previously.

The implementation includes a change to the `ReadHold` type, which now contains a sender closure, rather than a channel sender directly. The closure enables the construction of `ReadHold`s that communicate changes on a channel with an item type different from `(GlobalId, ChangeBatch)`.

### Motivation

  * This PR fixes a recognized bug.

Fixes https://github.com/MaterializeInc/database-issues/issues/8679

### Tips for reviewer

This PR required touching a lot more parts than I initially expected, so it ended up quite busy. I managed to split out sensible commits though, looking at those separately makes the reviewing experience much better!

The above mentioned bug has already been mitigated in https://github.com/MaterializeInc/materialize/pull/30109. This PR delivers an alternative fix that requires more restructuring but lets us end up in a saner state with fewer things moving concurrently. For example, with the other fix we have to always assume read hold changes for unknown collections are for future collections, which would make us miss bugs where we wrongly issue read hold changes for already dropped collections.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/storage-types/src/read_holds.rs [hunk 1]
```diff
@@ -83,18 +96,14 @@ impl<T: TimelyTimestamp> ReadHold<T> {
...
+    /// (different [GlobalId]).
     pub fn merge_assign(&mut self, mut other: ReadHold<T>) {
         assert_eq!(
             self.id, other.id,
             "can only merge ReadHolds for the same ID"
         );
-        assert!(
-            self.holds_tx.same_channel(&other.holds_tx),
-            "can only merge ReadHolds that come from the same issuer"
-        );
```
Comment (teskje) [inline src/storage-types/src/read_holds.rs:97]: If anybody feels strongly about keeping this check, we could make `ChangeTx` a custom trait, e.g.:

```
trait ChangeTx {
    fn send(&self, id: GlobalId, change: ChangeBatch) -> Result;
    fn same_recipient(&self, other: &Self) -> bool;
}
```

File: src/compute-client/src/controller.rs [hunk 2]
```diff
@@ -1143,7 +1112,12 @@ impl<T: ComputeControllerTimestamp> InstanceState<T> {
             since
         });
 
-        let hold = ReadHold::new(id, since, self.read_holds_tx.clone());
+        let client = self.client.clone();
+        let tx = Box::new(move |id, change| {
+            client.call(move |i| i.report_read_hold_change(id, change));
+            Ok(())
+        });
+        let hold = ReadHold::new(id, since, tx);
```
Comment (teskje) [inline src/compute-client/src/controller.rs:1120]: Note that `ReadHold`s issued by the compute controller now contain a copy of `instance::Client` and therefore keep the connected `Instance` alive. I don't see any immediate issue with that, but I might be missing something.

File: src/storage-types/src/read_holds.rs [hunk 3]
```diff
@@ -8,13 +8,21 @@
...
+use std::sync::Arc;
 
 use mz_repr::GlobalId;
 use thiserror::Error;
 use timely::progress::{Antichain, ChangeBatch, Timestamp as TimelyTimestamp};
 use timely::PartialOrder;
+use tokio::sync::mpsc::error::SendError;
 use tokio::sync::mpsc::UnboundedSender;
 
+type ChangeTx<T> = Box<
```
Comment (petrosagg) [inline src/storage-types/src/read_holds.rs:20]: Should we make this an `Arc` directly instead of having an arc wrapping a box?

File: src/storage-types/src/read_holds.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage-types/src/read_holds.rs:97]: You can do this without resorting to a trait. Effectively what you want to do is create a single prototypical `Arc<sender closure>` in the controller that gets cloned into all read holds that are given out.  Then this assert can be replaced with `Arc::ptr_eq`

File: src/compute-client/src/controller.rs [hunk 4]
```diff
@@ -1069,24 +1064,14 @@ struct InstanceState<T: ComputeControllerTimestamp> {
     client: instance::Client<T>,
     replicas: BTreeSet<ReplicaId>,
     collections: BTreeMap<GlobalId, Collection<T>>,
-    /// Sender for updates to collection read holds.
-    ///
-    /// Copies of this sender are given to [`ReadHold`]s that are created in
-    /// [`InstanceState::acquire_read_hold`].
-    read_holds_tx: mpsc::UnboundedSender<(GlobalId, ChangeBatch<T>)>,
```
Comment (petrosagg) [inline src/compute-client/src/controller.rs:1076]: (this would be where the prototypical sender closure would live)

File: src/storage-types/src/read_holds.rs (hunk 1 above)
Comment (teskje) [reply src/storage-types/src/read_holds.rs:97]: Ah, true. That would still allow people to create `ReadHold`s that write to the same channel but use different closures for it, but we can hope that CI would catch this.

File: src/storage-types/src/read_holds.rs (hunk 3 above)
Comment (teskje) [reply src/storage-types/src/read_holds.rs:20]: 💯

File: src/storage-types/src/read_holds.rs (hunk 1 above)
Comment (teskje) [reply src/storage-types/src/read_holds.rs]: Done

Comment (teskje) [reply src/storage-types/src/read_holds.rs]: Oh, actually I think we should probably not do this. With this scheme, two `ReadHold`s created with `with_channel` for the same channel won't be mergeable. So if you request two read holds for the same collection from the storage controller, then try to merge them, Mz will panic. CI suggests that we don't try to do such things today, but we might in edge-cases that are not covered by CI, in which case the assert would be a time bomb waiting to explode in production.

**Comment by teskje** [comment]
Finally had time to look at the test issues. Should be ready for another look now!

**Comment by jkosh44** [comment]
Looks like the adapter changes are minimal enough that you can proceed without an adapter approval.

File: src/compute-client/src/controller/instance.rs [hunk 5]
```diff
@@ -975,20 +976,35 @@ where
...
     #[mz_ore::instrument(level = "debug")]
-    pub fn check_empty(&mut self) {
+    pub fn shutdown(&mut self) {
+        // Taking the `command_rx` ensures that the [`Instance::run`] loop terminates.
+        let (_tx, rx) = mpsc::unbounded_channel();
+        let mut command_rx = std::mem::replace(&mut self.command_rx, rx);
+
+        // Apply all outstanding read hold changes. This might cause read hold downgrades to be
+        // added to `command_tx`, so we need to apply those in a loop.
+        while !self.read_holds_rx.is_empty() {
```
Comment (petrosagg) [inline src/compute-client/src/controller/instance.rs]: Why do we care to do that if we're shutting down the instance? Isn't it enough to assert that there are no unexpected commands in the command_rx buffer but otherwise ignore everything?

**Review by petrosagg — APPROVED** [review]
Looks good!

File: src/compute-client/src/controller/instance.rs (hunk 5 above)
Comment (teskje) [reply src/compute-client/src/controller/instance.rs]: The method comment hints at this:
```
    /// This method runs various assertions ensuring the instance state is empty. It exists to help
    /// us find bugs where the client drops a compute instance that still has replicas or
    /// collections installed, and later assumes that said replicas/collections still exists.
```

We sanity check below that the instance doesn't have any collections installed anymore, but for this we need to be sure all pending read holds on all collections have been dropped, so we can drop the collection state. It can be that compute's users have correctly dropped all read holds but we haven't applied those drops, or the transitive read hold drops resulting from them, yet. That's what this loop does.

You are right that we could just not do the assert and save us the effort instead. It is, in a way, a service to the adapter team. If we remove it, we lose the ability to detect bugs where the coordinator drops an instance but not collections installed on the instance, and then becomes confused. Not that I have a reason to assume these bugs are particularly likely though.

**Comment by teskje** [comment]
TFTR!

**Comment by bosconi** [comment]
bugbot run

File: src/compute-client/src/controller/instance.rs [hunk 6]
```diff
@@ -158,11 +167,20 @@ where
...
 
+        let read_hold_tx: read_holds::ChangeTx<_> = {
+            let command_tx = command_tx.clone();
+            Arc::new(move |id, change: ChangeBatch<_>| {
+                let cmd: Command<_> = {
+                    let change = change.clone();
+                    Box::new(move |i| i.apply_read_hold_change(id, change.clone()))
+                };
+                command_tx.send(cmd).map_err(|_| SendError((id, change)))
+            })
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/controller/instance.rs:181]: https://github.com/MaterializeInc/materialize/pull/35564
